File: bundle-plugin/src/main/scala/com/yahoo/container/plugin/bundle/TransformExportPackages.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.bundle

import java.io.File
import com.yahoo.container.plugin.osgi.ExportPackages.{Export, Parameter}
import com.yahoo.container.plugin.osgi.ExportPackages.Export

/**
 * @author tonytv
 */
object TransformExportPackages extends App {
  def replaceVersions(exports: List[Export], newVersion: String): List[Export] = {
    mapParameters(exports) { parameters =>
      parameters map replaceVersion(newVersion)
    }
  }

  def removeUses(exports: List[Export]): List[Export] = {
    mapParameters(exports) { parameters =>
      parameters filter {_.name != "uses"}
    }
  }

  def mapParameters(exports: List[Export])(f: List[Parameter] => List[Parameter]): List[Export] = {
    exports map { case Export(packageNames: List[String], parameters: List[Parameter]) =>
        Export(packageNames, f(parameters))
    }
  }

  private def replaceVersion(newVersion: String)(parameter: Parameter) = {
    parameter match {
      case Parameter("version", _) => Parameter("version", newVersion)
      case other => other
    }
  }

  def toExportPackageProperty(exports: List[Export]): String = {
    val exportPackages =
      exports map { case Export(packageNames: List[String], parameters: List[Parameter]) =>
        val parameterString = nameEqualsValue(parameters)
        (packageNames ++ parameterString) mkString ";"
      }

    exportPackages mkString ","
  }

  private def nameEqualsValue(parameters: List[Parameter]) = {
    parameters map { case Parameter(name, value) =>
      s"$name=${quote(value)}"
    }
  }

  def quote(s: String) = '"' + s + '"'
}

File: bundle-plugin/src/main/java/com/yahoo/container/plugin/bundle/TransformExportPackages.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.bundle;

import com.yahoo.container.plugin.osgi.ExportPackages.Export;
import com.yahoo.container.plugin.osgi.ExportPackages.Parameter;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Tony Vaagenes
 * @author ollivir
 */
public class TransformExportPackages {
    public static List<Export> replaceVersions(List<Export> exports, String newVersion) {
        List<Export> ret = new ArrayList<>();

        for (Export export : exports) {
            List<Parameter> newParams = new ArrayList<>();
            for (Parameter param : export.getParameters()) {
                if ("version".equals(param.getName())) {
                    newParams.add(new Parameter("version", newVersion));
                } else {
                    newParams.add(param);
                }
            }
            ret.add(new Export(export.getPackageNames(), newParams));
        }
        return ret;
    }

    public static List<Export> removeUses(List<Export> exports) {
        List<Export> ret = new ArrayList<>();

        for (Export export : exports) {
            List<Parameter> newParams = new ArrayList<>();
            for (Parameter param : export.getParameters()) {
                if ("uses".equals(param.getName()) == false) {
                    newParams.add(param);
                }
            }
            ret.add(new Export(export.getPackageNames(), newParams));
        }
        return ret;
    }

    public static String toExportPackageProperty(List<Export> exports) {
        return exports.stream().map(exp -> {
            String oneExport = String.join(";", exp.getPackageNames());
            if (exp.getParameters().size() > 0) {
                String paramString = exp.getParameters().stream().map(param -> param.getName() + "=" + quote(param.getValue())).collect(Collectors.joining(";"));
                oneExport += ";" + paramString;
            }
            return oneExport;
        }).collect(Collectors.joining(","));
    }

    public static String quote(String s) {
        return "\"" + s + "\"";
    }
}
